---
title: Typesafe Error Handling
description: How to intercept, handle or log errors inside oRPC
---

## Typed errors

```ts twoslash
import { createORPCReactQueryUtils } from '@orpc/react-query'
import { createProcedureClient, isDefinedError, os, safe } from '@orpc/server'

const createPost = os
  .errors({
    CONFLICT: {
      status: 409, // (optional) default CONFLICT is 409
      message: 'Conflict', // (optional)
      data: z.object({ // (optional)
        why: z.string(),
      }),
    },
    // ANY_CODE: { unknown code is allowed, with default status is 500
    //   status: 500,
    // },
  })
  .handler(({ errors }) => {
    throw errors.CONFLICT({ data: { why: 'some reason' } })

    // is the same as, but this is not typed
    throw new ORPCError({ code: 'CONFLICT', data: { why: 'some reason' } })

    // throw errors.ANY_CODE()
  })

const client = createProcedureClient({ procedure: createPost }) // or any kind of client

const [data, error, isDefined] = await safe(client({ title: 'title' }))

if (error) {
  if (isDefinedError(error)) { // or use isDefined const
    const data = error.data // { why: 'some reason' } full typed data
  }
  else {
    // any errors what not satisfy, or not defined in .errors
  }
}

/// example handle error for `@orpc/react-query`

import { useMutation } from '@tanstack/react-query'
import { z } from 'zod'

const utils = createORPCReactQueryUtils(client)

const mutation = useMutation(utils.mutationOptions({
  onError(error) {
    if (isDefinedError(error)) {
      const data = error.data // { why: 'some reason' } full typed data
    }
  },
}))
```

> **Note**: typesafe errors are coverage in every oRPC packages: `@orpc/client`, `@orpc/react-query`, `@orpc/vue-query`, `@orpc/vue-colada`, ...
> Please use `isDefinedError` or `safe` to typesafe handle errors. Export from `@orpc/contract`, `@orpc/server`, and `@orpc/client`.

## Handle errors

```ts twoslash
import { ORPCError, os } from '@orpc/server'

const ping = os
    .use(async ({ context, path, next }, input) => {
        try {
            const result = await next({})
            const _output = result.output // do something on success
            return result
        } catch (e) {
            // do something on error
            throw e
        } finally {
            // do something on finish
        }
    })
    .handler(({ input, context }) => {
        throw new ORPCError({
            code: 'NOT_FOUND',
            message: 'Not found',
            status: 404, // Optional: custom default behavior
            data: { something: 'include in the body and send to the client' } // pass data to the client
        })
    })
```

